ipam/store.go

```diff
@@ -109,50 +91,47 @@ func (a *Allocator) watchForChanges() error {
 	return nil
 }
 
-func (a *Allocator) readFromStore() error {
-	a.Lock()
-	store := a.store
-	a.Unlock()
-
-	if store == nil {
+func (cfg *PoolsConfig) writeToStore() error {
+	if cfg.ds == nil {
 		return nil
 	}
+	err := cfg.ds.PutObjectAtomic(cfg)
+	if err == datastore.ErrKeyModified {
+		return types.RetryErrorf("failed to perform atomic write (%v). retry might fix the error", err)
+	}
+	return err
+}
 
-	kvPair, err := a.store.KVStore().Get(datastore.Key(a.Key()...))
+func (cfg *PoolsConfig) readFromStore() error {
+	if cfg.ds == nil {
+		return nil
+	}
+	kvPair, err := cfg.ds.KVStore().Get(datastore.Key(cfg.Key()...))
```

Comment:
Can you please `ds.GetObject(key, cfg)` method instead of using the libkv `Get` method directly. With the `GetObject` you get `readFromKey` for free. You may still need it for watch but that is going away anyway.